### ISSUE TITLE
Lag in collaborative mode

### ISSUE THREAD

**Issue by ad1992** [excalidraw/excalidraw#1523]
If you start drawing diagrams and do frequent rotations / resizing/ moving shapes, it leads to at least 8-10 seconds and even goes to 15-20 seconds lag. Also, cursor sync takes good amount of bandwidth.

I tried adding debounce of around 30ms and it takes way less bandwidth compared to current state and I didn't notice any lag as well.
should we try that out?

Any other thoughts on optimizing?

**Comment by dwelle** [comment]
A duplicate of https://github.com/excalidraw/excalidraw/issues/1310 (?) --- but we can rename this issue and track it as "add debounce to syncing".

The worry about adding debounce is convergence issues. Native debounce drops intermediate updates, which will often be significant. We'd need to implement update batching & merging, which is non-trivial.

**Comment by ad1992** [comment]
Hmm, so to start with I was trying the below optimizations.
* Debounce mouse events
* Instead of calling broadcastScene from componentDidUpdate, I created a setInterval and started calling broadcastScene with an acceptable interval (tried with 30ms)

If needed I can push this to branch so the team can try it and we can then decide if this looks fine and then as you said I can pick batching and merging,

**Comment by dwelle** [comment]
> I created a setInterval and started calling broadcastScene with an acceptable interval (tried with 30ms)

Not sure if I understand, but adding a `setInterval` doesn't sound like a good idea. I'd just set up a debounce-like handler as you suggested.

But for `broadcastScene` you'd really need to come up with a queue & merging scheme in order to be accepted I'm afraid.

I personally am not against it, but you'll probably get pushback unless it's implemented in a way that 1) guarantees synchronization convergence (and without introducing some crazy complexity on top), and 2) doesn't deteriorate UX for people who aren't already suffering from the lag as you and me are -- which seems to be most people in the US.

**Comment by ad1992** [comment]
Sure I will look into queuing and merging schemes for broadcast scene. For pointer debounce actually helps and takes way less bandwidth now

**Comment by dwelle** [comment]
Come to think of it, since we're syncing whole scene, there shouldn't really be any convergence issues. Would need to review.

Also, my brain failed again --- instead of debounce, shouldn't we use throttle? Will need to check your mouse PR to see how it behaves with debounce.

**Comment by ad1992** [comment]
Hmm now that I think of throttle might be better so that once in every 30ms for sure the sync happens instead of waiting for 30ms after mouse movements stop. 

Also since whole scene gets updated as you mentioned, I think throttling the broadcast scene will also help, I will try that too.

In the meanwhile, you can check once the mouse movements with debounce but I agree throttle will be better here.

**Comment by vjeux** [comment]
I don't think that blindly adding a debounce/throttle is the right approach. The fact that we get real time events is a feature, it makes the collaboration feel real time and is making the experience better. Most of the time it's actually working fine for me.

I don't believe that blindly adding a debounce/throttle is the right thing to do here. Some of the next steps we could do:

1) Your profile says that you live in Bangalore, India. The server is in the US. So one thing we could do is to start having more than one server and use one close to where the users are.

2) Right now, we're sending the entire scene on every update. We know that this is suboptimal. Making it send only the delta would be better. @petehunt has ideas on how to make that happen.

3) What is happening in this case is that the upload bandwidth is smaller than what we're trying to send, so it serializes all the sends, hence the lag (I assume, we'd need to figure if it's actually the real reason). We could avoid blindingly send all the updates but instead send an update, while it's sending, put everything in a queue, wait for the ack, and drop all the messages but the last one that we were going to send and send it.

**Comment by ad1992** [comment]
@vjeux some updates might get missed when throttling the broadcast scene and sending the delta as you suggested would be great. but for mouse events, I think we should add throttle, what are cons you see. in that?

**Comment by vjeux** [comment]
If we're able to see the mouse updating real time at 60fps (which has been my experience with the current collaboration mode), why would we artificially make it worse?

**Comment by ad1992** [comment]
I get your concern @vjeux, my only concern is syncing the cursors is taking a good amount of bandwidth so to reduce the traffic I suggested if we can add throttle/debounce with some expectable duration.

**Comment by vjeux** [comment]
@ad1992 can you be more specific? What kind of bandwidth are we talking about here. And what is the consequence, are you worried about metered internet plans?

**Comment by ad1992** [comment]
@vjeux the frequency at which the data is getting transferred(i am checking the web socket in chrome console), is very high since its real-time which might actually start impacting user experience (slowness in collaborative mode). This was observed when testing with 3-4 users and the experience improved when turning off the cursor sync.
So I was looking for some optimizations when syncing cursor. And I am not saying to sync every x ms, instead throttle/debounce the handlers itself is what I meant as in the pr, so there won't be any data missing in sync.

As for the lag of scene, the update is mostly observed when too much of activity is done continuously example rotation shapes very fast (unreal scenario)

**Comment by dwelle** [comment]
My idea is to throttle at 16ms (to target 60FPS), but we'd need to first ascertain that we're already sending updates more frequently than that.

We could also batch mouse updates with scene updates.

But as I mentioned, all this is adding complexity. Also note that in the long-term, we're planning to move to sending deltas only (not whole scene), which will improve the perf. And when we do so, any naive throttling we'd be doing (as part of this issue/PR) would need to either be reverted, or improved to ensure we merge those deltas.

**Comment by vjeux** [comment]
16ms is the refresh rate of the screen and browsers don't send more than one event per frame, so this would not achieve anything more than introducing an artificial delay.

**Comment by dwelle** [comment]
Oh, I didn't know 60FPS is a hard limit. In that case, never mind. But, there *is* a way for browsers to execute multiple tasks in a single frame, no? E.g., when you schedule several tasks via `requestAnimationFrame`, they'll tend to be executed in the same frame (either at the end of current frame, or at the start of next frame) --- not that it's much relevant to what we're solving here.

**Comment by silverark** [comment]
Just as a headsup: I've just been looking at adding a throttle for `pointermove` as I was seeing 300 mouse updates per second. It appears that there is a bug in Chrome where there are more pointer updates sent than the render step when you open dev tools. I was obviously using devtools to test! I couldn't work out why one collaborator was slow and the other was really fast! 

https://bugs.chromium.org/p/chromium/issues/detail?id=992954

**Comment by dwelle** [comment]
Interesting catch!
